hnix-store-core/src/System/Nix/Internal/Hash.hs

```diff
@@ -0,0 +1,152 @@
+                         then Just (Digest bs)
+                         else Nothing
+
+  initialize        :: AlgoCtx a
+  update            :: AlgoCtx a -> BS.ByteString -> AlgoCtx a
+  finalize          :: AlgoCtx a -> Digest a
+
+  toBase32          :: Digest a -> T.Text
+  -- ^ The base32 representation of a @Digest a@ is a text
+  --   encoding. TODO - reference thesis spec?
+
+
+-- | The cryptographic hash of of a strict bytestring, where hash
+--   algorithm is chosen by the type of the digest
+--   For example:
+--   > let d = hash "Hello, sha-256!" :: Digest SHA256
+--   or
+--   > :set -XTypeApplications
+--   > let d = hash @SHA256 "Hello, sha-256!"
+hash :: forall a.HasDigest a => BS.ByteString -> Digest a
+hash bs =
+  finalize $ update @a (initialize @a) bs
+
+-- | The cryptographic hash of a lazy bytestring. Use is the same
+--   as for @hash@. This runs in constant space, but forces the
+--   entire bytestring
+hashLazy :: forall a.HasDigest a => BSL.ByteString -> Digest a
+hashLazy bsl =
+  finalize $ foldl' (update @a) (initialize @a) (BSL.toChunks bsl)
+
+
+instance HasDigest MD5 where
+  type AlgoCtx 'MD5 = MD5.Ctx
+  hashLength _ = 16
+  initialize = MD5.init
+  update = MD5.update
+  finalize = Digest . MD5.finalize
+  toBase32 (Digest bs) = T.decodeUtf8 $ printHashBytes32 bs
+
+instance HasDigest 'SHA1 where
+  type AlgoCtx SHA1 = SHA1.Ctx
+  hashLength _ = 20
+  initialize = SHA1.init
+  update = SHA1.update
+  finalize = Digest . SHA1.finalize
+  toBase32 (Digest bs) = T.decodeUtf8 $ printHashBytes32 bs
+
+instance HasDigest 'SHA256 where
+  type AlgoCtx SHA256 = SHA256.Ctx
+  hashLength _ = 32
+  initialize = SHA256.init
+  update = SHA256.update
+  finalize = Digest . SHA256.finalize
+  toBase32 (Digest bs) =
+    T.decodeUtf8 $ printHashBytes32 $ truncateDigest 20 bs
+
+-- | A raw hash digest, with a type-level tag
+data Digest (a :: HashAlgorithm) = Digest
+  { digestBytes :: BS.ByteString
+    -- ^ The bytestring in a Digest is an opaque string of bytes,
+    --   not some particular text encoding.
+  } deriving (Show, Eq, Ord)
+
+instance DataHashable.Hashable (Digest a) where
+  hashWithSalt a (Digest bs) = DataHashable.hashWithSalt a bs
+
+
+-- TODO: Test this
+printHashBytes32 :: BS.ByteString -> BSC.ByteString
+printHashBytes32 c = BSC.pack $ concatMap char32 [nChar - 1, nChar - 2 .. 0]
+  where
+    -- The base32 encoding is 8/5's as long as the base256 digest
+    nChar = fromIntegral $ BS.length c * 8 `div` 5
+
+    char32 :: Integer -> [Char]
+    char32 i = [digits32 V.! digitInd]
+      where
+        byte j   = BS.index c (fromIntegral j)
+        digitInd = fromIntegral $
+                   sum [fromIntegral (byte j) * (256^j)
+                       | j <- [0 .. BS.length c - 1]]
+                   `div` (32^i)
+                   `mod` 32
+
+
+-- TODO Test this, too
+truncateDigest :: Int -> BS.ByteString -> BS.ByteString
+truncateDigest n c = BS.pack $ map truncOutputByte [0.. n-1]
```

Comment:
Why isn't this just `BS.take`?

Reply:
This is a straight (and inefficient!) translation of the thesis. Because `n < |c|` (truncated length is less than the input bytestring length), some positions in the input bytestring get to impact the truncated version twice.

A more straightforward implementation would be: 
```
let layer1  = BS.take n bs
    -- leading bits that aren't truncated
    layer2 = BS.drop n bs <> BS.replicate (BS.length bs - n) 0
    -- ^ trailing bits, right padded
    -- v xor the two together for "truncated" digest
in xor layer1 layer2
``` (if xor could work bitwise over a bytestring)

Reply:
Wait, seriously? It's not just a naive truncation? Why???

Reply:
```c++
    Hash h;
    h.hashSize = newSize;
    for (unsigned int i = 0; i < hash.hashSize; ++i)
        h.hash[i % newSize] ^= hash.hash[i];
    return h;
```
OK then